src/bootstrap/step.rs

```diff
@@ -838,14 +838,22 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
             } else {
                 &self.build.config.target
             };
-            // If --target was specified but --host wasn't specified, don't run
-            // any host-only tests
+            // Determine the actual targets participating in this rule.
             let arr = if rule.host {
-                if self.build.flags.target.len() > 0 &&
-                   self.build.flags.host.len() == 0 {
-                    &hosts[..0]
+                if kind == Kind::Dist {
+                    // Keep the full projection from build triple to targets
+                    // for the dist steps, now that the hosts array is truncated
+                    // to avoid duplication of work.
+                    targets
```

Comment:
I think we'll still want to draw from the array of hosts here rather than use the array of targets. If `rule.host` is true then we don't want to try to package compilers for only-target targets. For example if you configure rustbuild with `--build=A --host=B --target=C`, then we should see:

* std for A, B, C
* rustc for A, B
* rust-docs for A, B, C
* etc

I *think* this will cause rustc to be package for A, B, and C, which'd cause failures.

I believe one way to solve this though would be to leave the logic as-is from before but shadow the outer `hosts` variable with the same definition above but minus the `Dist` logic. Does that make sense?